### ISSUE TITLE
Can't link debug build without UBSAN active

### ISSUE THREAD

**Issue by eaigner** [microsoft/appcenter-sdk-apple#2288]
### **Description**

When trying to run a debug build, and the "Thread Sanitizer" (or any scheme option that links UBSAN) is not enabled, the linker refuses to link with missing symbols:

~~~
Undefined symbols for architecture arm64:
  "___ubsan_handle_add_overflow", referenced from:
      -[MSACAnalyticsAuthenticationProvider checkTokenExpiry] in AppCenterAnalytics.o
      -[MSACHttpCall startRetryTimerWithStatusCode:retryAfter:event:] in AppCenter.o
      ___74+[MSACDBStorage executeSelectionQuery:inOpenedDatabase:result:withValues:]_block_invoke in AppCenter.o
      -[MSACDBStorage setMaxStorageSize:completionHandler:] in AppCenter.o
      ___46-[MSACLogDBStorage saveLog:withGroupId:flags:]_block_invoke in AppCenter.o
      -[MSACStorageBindableArray bindAllValuesWithStatement:inOpenedDatabase:] in AppCenter.o
      ___72-[MSACStorageBindableArray bindAllValuesWithStatement:inOpenedDatabase:]_block_invoke in AppCenter.o
      ...
  "___ubsan_handle_builtin_unreachable", referenced from:
      ___cxa_throw in AppCenterCrashes.o
      MSACCrashesUncaughtCXXTerminateHandler() in AppCenterCrashes.o
      std::__1::vector<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*), std::__1::allocator<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*)> >::__recommend(unsigned long) const in AppCenterCrashes.o
      std::__1::allocator<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*)>::allocate(unsigned long, void const*) in AppCenterCrashes.o
      uncaught_cxx_exception_handler(MSACCrashesUncaughtCXXExceptionInfo const*) in AppCenterCrashes.o
      std::__1::enable_if<__is_cpp17_forward_iterator<char const*>::value, void>::type std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init<char const*>(char const*, char const*) in AppCenterCrashes.o
      std::__1::allocator<char>::allocate(unsigned long, void const*) in AppCenterCrashes.o
      ...
  "___ubsan_handle_divrem_overflow", referenced from:
      -[MSACDeviceTracker timeZoneOffset:] in AppCenter.o
      -[MSACAbstractLog convertTimeZoneOffsetToISO8601:] in AppCenter.o
      -[MSACDBStorage executeQueryUsingBlock:] in AppCenter.o
      -[MSACDBStorage setMaxStorageSize:completionHandler:] in AppCenter.o
      std::__1::vector<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*), std::__1::allocator<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*)> >::__recommend(unsigned long) const in AppCenterCrashes.o
      _protobuf_c_message_unpack in CrashReporter.o
      _int_range_lookup in CrashReporter.o
      ...
  "___ubsan_handle_float_cast_overflow", referenced from:
      -[MSACAnalyticsAuthenticationProvider checkTokenExpiry] in AppCenterAnalytics.o
      -[MSACChannelUnitDefault resolveFlushInterval] in AppCenter.o
      -[MSACDeviceTracker screenSize] in AppCenter.o
      -[MSACHttpCall delayForRetryCount:] in AppCenter.o
      -[MSACOneCollectorIngestion getHeadersWithData:eTag:] in AppCenter.o
      +[MSACCompression compressData:] in AppCenter.o
      +[MSACDispatcherUtil dispatchSyncWithTimeout:onQueue:withBlock:] in AppCenter.o
      ...
  "___ubsan_handle_load_invalid_value", referenced from:
      -[MSACSessionTracker hasSessionTimedOut] in AppCenterAnalytics.o
      -[MSACSessionTracker isStarted] in AppCenterAnalytics.o
      -[MSACSessionTracker setStarted:] in AppCenterAnalytics.o
      -[NSViewController(PageViewLogging) ms_viewWillAppear:] in AppCenterAnalytics.o
      -[MSACAnalytics startWithChannelGroup:appSecret:transmissionTargetToken:fromApplication:] in AppCenterAnalytics.o
      -[MSACAnalytics setEnabled:] in AppCenterAnalytics.o
      -[MSACAnalytics applyEnabledState:] in AppCenterAnalytics.o
      ...
  "___ubsan_handle_mul_overflow", referenced from:
      -[MSACDBStorage setMaxStorageSize:completionHandler:] in AppCenter.o
      ___46-[MSACLogDBStorage saveLog:withGroupId:flags:]_block_invoke in AppCenter.o
      _plcrash_async_address_apply_offset in CrashReporter.o
      _plcrash_async_cfe_register_decode in CrashReporter.o
      plcrash::async::dwarf_cfa_state<unsigned int, int>::eval_program(plcrash_async_mobject*, unsigned int, unsigned int, plcrash::async::plcrash_async_dwarf_cie_info*, plcrash::async::gnu_ehptr_reader<unsigned int>*, plcrash_async_byteorder const*, unsigned long long, long long, unsigned long long) in CrashReporter.o
      plcrash::async::dwarf_cfa_state<unsigned long long, long long>::eval_program(plcrash_async_mobject*, unsigned long long, unsigned long long, plcrash::async::plcrash_async_dwarf_cie_info*, plcrash::async::gnu_ehptr_reader<unsigned long long>*, plcrash_async_byteorder const*, unsigned long long, long long, unsigned long long) in CrashReporter.o
      _plcrash_async_mobject_verify_local_pointer in CrashReporter.o
      ...
  "___ubsan_handle_negate_overflow", referenced from:
      plcrash_error_t plcrash::async::plcrash_async_dwarf_expression_eval<unsigned int, int>(plcrash_async_mobject*, unsigned int, plcrash_async_thread_state const*, plcrash_async_byteorder const*, unsigned long long, long long, unsigned long long, unsigned int*, unsigned long, unsigned int*) in CrashReporter.o
      plcrash_error_t plcrash::async::plcrash_async_dwarf_expression_eval<unsigned long long, long long>(plcrash_async_mobject*, unsigned int, plcrash_async_thread_state const*, plcrash_async_byteorder const*, unsigned long long, long long, unsigned long long, unsigned long long*, unsigned long, unsigned long long*) in CrashReporter.o
      _plcrash_nasync_macho_init in CrashReporter.o
      _zigzag32 in CrashReporter.o
      _zigzag64 in CrashReporter.o
  "___ubsan_handle_nonnull_arg", referenced from:
      +[NSViewController(PageViewLogging) swizzleViewWillAppear] in AppCenterAnalytics.o
      +[MSACAnalytics sharedInstance] in AppCenterAnalytics.o
      -[MSACAnalytics applyEnabledState:] in AppCenterAnalytics.o
      -[MSACChannelGroupDefault addChannelUnitWithConfiguration:withIngestion:] in AppCenter.o
      -[MSACChannelGroupDefault setMaxStorageSize:completionHandler:] in AppCenter.o
      -[MSACChannelUnitDefault addDelegate:] in AppCenter.o
      -[MSACChannelUnitDefault removeDelegate:] in AppCenter.o
      ...
  "___ubsan_handle_out_of_bounds", referenced from:
      -[MSACHttpIngestion setBaseURL:] in AppCenter.o
      ___35+[MSACAppCenter isDebuggerAttached]_block_invoke in AppCenter.o
      ___cxa_throw in AppCenterCrashes.o
      MSACCrashesUncaughtCXXTerminateHandler() in AppCenterCrashes.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__get_short_pointer() const in AppCenterCrashes.o
      std::__1::array<MSACCrashesBufferedLog, 60ul>::operator[](unsigned long) in AppCenterCrashes.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__zero() in AppCenterCrashes.o
      ...
  "___ubsan_handle_pointer_overflow", referenced from:
      -[MSACHttpIngestion setBaseURL:] in AppCenter.o
      +[MSACCompression compressData:] in AppCenter.o
      +[NSObject(MSACUtility_StringFormatting) sha256:] in AppCenter.o
      ___35+[MSACAppCenter isDebuggerAttached]_block_invoke in AppCenter.o
      ___cxa_throw in AppCenterCrashes.o
      MSACCrashesUncaughtCXXTerminateHandler() in AppCenterCrashes.o
      std::__1::vector<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*), std::__1::allocator<void (*)(MSACCrashesUncaughtCXXExceptionInfo const*)> >::erase(std::__1::__wrap_iter<void (* const*)(MSACCrashesUncaughtCXXExceptionInfo const*)>) in AppCenterCrashes.o
      ...
  "___ubsan_handle_shift_out_of_bounds", referenced from:
      +[MSACErrorLogFormatter formatAddress:is64bit:] in AppCenterCrashes.o
      _protobuf_c_message_unpack in CrashReporter.o
      _parse_tag_and_wiretype in CrashReporter.o
      _scan_length_prefixed_data in CrashReporter.o
      _get_tag_size in CrashReporter.o
      _int32_size in CrashReporter.o
      _uint32_size in CrashReporter.o
      ...
  "___ubsan_handle_sub_overflow", referenced from:
      MSACCrashesUncaughtCXXTerminateHandler() in AppCenterCrashes.o
      _int_range_lookup in CrashReporter.o
      plcrash_error_t plcrash::async::plcrash_async_dwarf_expression_eval<unsigned int, int>(plcrash_async_mobject*, unsigned int, plcrash_async_thread_state const*, plcrash_async_byteorder const*, unsigned long long, long long, unsigned long long, unsigned int*, unsigned long, unsigned int*) in CrashReporter.o
      plcrash_error_t plcrash::async::plcrash_async_dwarf_expression_eval<unsigned long long, long long>(plcrash_async_mobject*, unsigned int, plcrash_async_thread_state const*, plcrash_async_byteorder const*, unsigned long long, long long, unsigned long long, unsigned long long*, unsigned long, unsigned long long*) in CrashReporter.o
  "___ubsan_handle_type_mismatch_v1", referenced from:
      -[MSACSessionTracker init] in AppCenterAnalytics.o
      -[MSACSessionTracker hasSessionTimedOut] in AppCenterAnalytics.o
      -[MSACSessionTracker setSessionTimeout:] in AppCenterAnalytics.o
      -[MSACSessionTracker setStarted:] in AppCenterAnalytics.o
      +[NSViewController(PageViewLogging) swizzleViewWillAppear] in AppCenterAnalytics.o
      -[MSACAnalytics init] in AppCenterAnalytics.o
      +[MSACAnalytics sharedInstance] in AppCenterAnalytics.o
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
~~~

I tried rebuilding the packages and clearing the cache several times, as well as clearing the build folder, with no success.

### **Repro Steps**

Please list the steps used to reproduce your issue.

1. Disable "Thread Sanitizer" in the Scheme
2. AppCenter fails to link

### **Details**

1. Which SDK version are you using?
    - 4.1.1
2. Which OS version did you experience the issue on?
    - macOS 11
3. Which Xcode version did you build the app with?
    - Version 12.4 (12D4e)
5. Which Cocoapods version are you using (run `pod --version`)?
    - none, Cocoapods suck.
6. What device version did you see this error on?  Were you using an emulator or a physical device?
    - mac Mini M1
7. What language are you using?
    - [ ] Objective C
    - [x] Swift
8. What third party libraries are you using?
    - None

**Comment by russelarms** [comment]
Hi @eaigner, thanks for reaching out!
What did you use to install the SDK?
I've checked the SDK on mac mini m1, Xcode 12.4, and SwiftPM for a macOS project, works fine for me.
Would you mind sharing a sample app for the problem?

**Comment by eaigner** [comment]
SwiftPM. If I inspect the generated object files generated when building the swift package, it also says there are UBSAN symbols in there.

~~~
% nm AppCenterAnalytics.o | grep  __ubsan
                 U ___ubsan_handle_add_overflow
                 U ___ubsan_handle_float_cast_overflow
                 U ___ubsan_handle_load_invalid_value
                 U ___ubsan_handle_nonnull_arg
                 U ___ubsan_handle_type_mismatch_v1
~~~

**Comment by russelarms** [comment]
@eaigner It seems like a bug inside Xcode/Xcode toolchain.
If I enable ubsan/tsan in Diagnostics options there are some injected code added to the object files:
```
nm AppCenter.o | grep ubsan
                 U ___ubsan_handle_add_overflow
                 U ___ubsan_handle_divrem_overflow
                 U ___ubsan_handle_float_cast_overflow
                 U ___ubsan_handle_load_invalid_value
                 U ___ubsan_handle_mul_overflow
                 U ___ubsan_handle_nonnull_arg
                 U ___ubsan_handle_out_of_bounds
                 U ___ubsan_handle_pointer_overflow
                 U ___ubsan_handle_type_mismatch_v1
```
```
nm AppCenter.o | grep san  
                 U ___tsan_atomic8_load
                 U ___tsan_atomic8_store
                 U ___tsan_func_entry
                 U ___tsan_func_exit
                 U ___tsan_ignore_thread_begin
                 U ___tsan_ignore_thread_end
                 U ___tsan_init
                 U ___tsan_read1
                 U ___tsan_read4
                 U ___tsan_read8
                 U ___tsan_write1
                 U ___tsan_write4
                 U ___tsan_write8
```
But after I disable it, it is completely removed and the app compiles smoothly. 
There's no output from `nm AppCenter.o | grep san`.

Probably worth trying with a different version of Xcode/Xcode CLT.

**Comment by AnastasiaKubova** [comment]
We haven’t heard from you in a while so I’m closing this issue but feel free to reopen this issue if you have additional questions.

**Comment by eaigner** [comment]
Still not working with Xcode Version 12.5 (12E262).

To be clear, this must be an AppCenter issue, since it occurs with no other SPM package, just AppCenter.

**Comment by AnastasiaKubova** [comment]
Hi! I checked this behavior with Xcode 12.5 following the next steps:
- Create macOS project
- Add App Center SDK via SPM
- Disable "Thread Sanitizer" in the scheme and run the app.
- Enable "Thread Sanitizer" in the scheme and run the app.

In both cases, all work well for me. It probably related to the specific settings of your project. 
Could you please try to reproduce this issue on an empty project and let me know about the results?

**Comment by eaigner** [comment]
OK I was able to solve this. I did some extensive testing with sample projects, but was unable to reproduce it. However I noticed the `enableUBSanitizer` flag can be set to `YES`, despite Xcode marking the `Undefined Behavior Sanitizer` option as disabled.

So what fixed it was to manually edit the schemes XML to remove the `enableUBSanitizer = "YES"` line.

While this does indeed seem to be an Xcode bug, it still weirdly only occurs with the AppCenter package.
